internal/expr/prepare.go

```diff
@@ -0,0 +1,86 @@
+package expr
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+)
+
+// PreparedExpr contains an SQL expression that is ready for execution.
+type PreparedExpr struct {
+	inputs []*inputPart
+	SQL    string
+}
+
+type typeNameToInfo map[string]*info
+
+func getKeys(m map[string]*info) []string {
+	i := 0
+	keys := make([]string, len(m))
+	for k := range m {
+		keys[i] = k
+		i++
+	}
+	return keys
```

Comment:
Are tests passing consistently?  Iterating over the map is unstable, so this needs a `sort.Strings` call before return.

Reply:
I clearly hadn't run it enough times! Changed.